driver-core/src/main/com/mongodb/operation/OperationHelper.java

```diff
@@ -414,6 +420,35 @@ static boolean serverIsAtLeastVersion(final ConnectionDescription description, f
         }
     }
 
+    static <T> T withReleasableConnection(final WriteBinding binding, final CallableWithConnectionAndSource<T> callable) {
+        ConnectionSource source = binding.getWriteConnectionSource();
+        try {
+            return callable.call(source, source.getConnection());
+        } finally {
+            source.release();
+        }
+    }
+
+    static <T> T withReleasableConnection(final WriteBinding binding, final Throwable connectionException,
```

Comment:
I think this can be simplified now after the latest refactorings.  Unlike in the first attempt where the connection has to be released before the second attempt, there is no need to do that here, and this method can be responsible for releasing the connection.

It also doesn't need to take a Throwable, as all callers are passing a MongoException.

Reply:
I've left this as is to keep it DRY - as its also used by MixedBulkWriteOperation that does need to keep hold of the connection.  In the future as we add more retryable operations we could add a `withConnection` method that takes a MongoException to throw if connecting fails.

Reply:
Ah, ok.